// connect-store/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
function enhanceStore(store) {
    var _store = store;
    if (_store.__$__enhanced) {
        return _store;
    }
    _store.__$__watchers = [];
    _store.__$__refCount = 0;
    var fields = Object.getOwnPropertyNames(_store).filter(function (k) { return !/^_/.test(k); });
    var properties = {};
    fields.forEach(function (field) {
        var innerField = "__$$__" + field;
        _store[innerField] = _store[field];
        var setter = _store.__lookupSetter__(field);
        properties[field] = {
            set: function (value) {
                if (setter) {
                    setter.bind(this)(value);
                }
                this[innerField] = value;
                this.__$__watchers.forEach(function (watcher) { return watcher(field, value); });
            },
            get: function () {
                return this[innerField];
            },
        };
    });
    properties.__$__state = {
        get: function () {
            return fields.reduce(function (memo, field) {
                memo[field] = _store[field];
                return memo;
            }, {});
        },
    };
    Object.defineProperties(_store, properties);
    _store.__$__enhanced = true;
    return _store;
}
var SyncQueue = (function () {
    function SyncQueue(wrapper, flushKeys) {
        this._queue = [];
        this._activated = false;
        this._shouldRelease = false;
        this._wrapper = wrapper;
        this._flushKeys = flushKeys;
    }
    SyncQueue.prototype.release = function () {
        this._shouldRelease = true;
    };
    SyncQueue.prototype.push = function (state) {
        if (this._flushKeys.length > 0 && this._flushKeys.includes(Object.keys(state)[0])) {
            this._wrapper.setState(state);
            return;
        }
        this._queue.push(state);
        this._activate();
    };
    SyncQueue.prototype._activate = function () {
        var _this = this;
        if (this._activated) {
            return;
        }
        this._activated = true;
        setImmediate(function () {
            if (_this._shouldRelease) {
                return;
            }
            _this._wrapper.setState(Object.assign.apply(Object, [{}].concat(_this._queue)));
            _this._queue = [];
            _this._activated = false;
        });
    };
    return SyncQueue;
}());
var Connector = (function () {
    function Connector(View) {
        this._dataSources = [];
        this._delegates = [];
        this._View = View;
    }
    Connector.prototype.connect = function (store) {
        this.dataSource(store);
        this.delegate(store);
        return this;
    };
    Connector.prototype.dataSource = function (store) {
        var enhancedStore = enhanceStore(store);
        if (this._dataSources.includes(enhancedStore)) {
            return this;
        }
        this._dataSources.push(enhancedStore);
        enhancedStore.__$__refCount += 1;
        return this;
    };
    Connector.prototype.delegate = function (delegate) {
        var enhancedStore = enhanceStore(delegate);
        if (this._delegates.includes(enhancedStore)) {
            return this;
        }
        this._delegates.push(enhancedStore);
        enhancedStore.__$__refCount += 1;
        return this;
    };
    Connector.prototype.render = function (staticProps, flushKeys) {
        if (flushKeys === void 0) { flushKeys = []; }
        var connector = this;
        var Wrapper = (function (_super) {
            __extends(Wrapper, _super);
            function Wrapper(props) {
                var _this = _super.call(this, props) || this;
                _this.queue = new SyncQueue(_this, flushKeys);
                _this.watcher = function (field, value) {
                    var _a;
                    _this.queue.push((_a = {},
                        _a[field] = value,
                        _a));
                };
                _this.state = Object.assign.apply(Object, [{}].concat(connector._dataSources.map(function (dataSource) {
                    dataSource.__$__watchers.push(_this.watcher);
                    return dataSource.__$__state;
                })));
                _this.delegate = new Proxy({}, {
                    get: function (target, p, receiver) {
                        if (/^_/.test(p)) {
                            throw new Error("Do not bind private methods of store. You may bind method '" + p + "'");
                        }
                        return function () {
                            var args = [];
                            for (var _i = 0; _i < arguments.length; _i++) {
                                args[_i] = arguments[_i];
                            }
                            connector._delegates.forEach(function (delegate) {
                                try {
                                    if (delegate[p] && typeof delegate[p] === 'function') {
                                        delegate[p].apply(delegate, args);
                                    }
                                }
                                catch (e) {
                                    console.error(e);
                                }
                            });
                        };
                    },
                });
                return _this;
            }
            Wrapper.prototype.componentWillUnmount = function () {
                var _this = this;
                var checkRefCount = function (store) {
                    if (store.__$__refCount <= 0) {
                        if (store.__idle && typeof store.__idle === 'function') {
                            store.__idle();
                        }
                    }
                };
                connector._dataSources.forEach(function (dataSource) {
                    dataSource.__$__watchers.splice(dataSource.__$__watchers.indexOf(_this.watcher), 1);
                    dataSource.__$__refCount -= 1;
                    checkRefCount(dataSource);
                });
                connector._delegates.forEach(function (delegate) {
                    delegate.__$__refCount -= 1;
                    checkRefCount(delegate);
                });
                connector._delegates = [];
                this.queue.release();
            };
            Wrapper.prototype.render = function () {
                var View = connector._View;
                return React.createElement(View, __assign({}, this.props, { dataSource: this.state, delegate: this.delegate }));
            };
            return Wrapper;
        }(React.PureComponent));
        return (React.createElement(Wrapper, __assign({}, staticProps)));
    };
    return Connector;
}());
exports.default = Connector;
